lib/winapi/findfile.c

```diff
@@ -0,0 +1,129 @@
+#include <string.h>
+#include <assert.h>
+#include <stdbool.h>
+#include <winbase.h>
+#include <xboxkrnl/xboxkrnl.h>
+#include <hal/fileio.h>
+#include <hal/winerror.h>
+#include <fileapi.h>
+
+struct FileInfo {
+    FILE_DIRECTORY_INFORMATION dirInfo;
+    char filename[MAX_PATH];
+};
+
+static void dirtofind (FILE_DIRECTORY_INFORMATION *dirInfo, LPWIN32_FIND_DATAA lpFindFileData)
+{
+    lpFindFileData->dwFileAttributes = dirInfo->FileAttributes;
+    lpFindFileData->ftCreationTime.dwLowDateTime = dirInfo->CreationTime.LowPart;
+    lpFindFileData->ftCreationTime.dwHighDateTime = dirInfo->CreationTime.HighPart;
+    lpFindFileData->ftLastAccessTime.dwLowDateTime  = dirInfo->LastAccessTime.LowPart;
+    lpFindFileData->ftLastAccessTime.dwHighDateTime = dirInfo->LastAccessTime.HighPart;
+    lpFindFileData->ftLastWriteTime.dwLowDateTime  = dirInfo->LastWriteTime.LowPart;
+    lpFindFileData->ftLastWriteTime.dwHighDateTime = dirInfo->LastWriteTime.HighPart;
+    lpFindFileData->nFileSizeHigh = dirInfo->EndOfFile.HighPart;
+    lpFindFileData->nFileSizeLow = dirInfo->EndOfFile.LowPart;
+    memcpy(lpFindFileData->cFileName, dirInfo->FileName, dirInfo->FileNameLength);
+    lpFindFileData->cFileName[dirInfo->FileNameLength] = '\0';
+    lpFindFileData->cAlternateFileName[0] = '\0';
+}
+
+HANDLE FindFirstFileA (LPCSTR lpFileName, LPWIN32_FIND_DATAA lpFindFileData)
+{
+    NTSTATUS status;
+    ANSI_STRING dirPath;
+    ANSI_STRING mask;
+    IO_STATUS_BLOCK ioStatusBlock;
+    OBJECT_ATTRIBUTES attributes;
+    struct FileInfo fileInformation;
+    HANDLE handle;
+    size_t maskOffset;
+    char fullPath[MAX_PATH+29];
```

Comment:
Where does this 29 come from?

Reply:
29 is the maximum amount of characters added by `XConvertDOSFilenameToXBOX`. It's added here to make sure the buffer is always large enough.

Reply:
That's a bit strange (I'd rather add a safer value, like 64). But okay.

I assume `XConvertDOSFilenameToXBOX` will be gone soon enough anyway.